### ISSUE TITLE
[data grid] Implement Row reorder

### ISSUE THREAD

**Issue by oliviertassinari** [mui/mui-x#206]
### Benchmark

- https://ag-grid.com/javascript-grid-row-dragging/
- https://www.telerik.com/kendo-react-ui/components/grid/advanced-features/row-reordering/
- https://ant.design/components/table/#components-table-demo-drag-sorting
- https://ej2.syncfusion.com/react/demos/#/material/grid/drag-drop-within-grid
- https://handsontable.com/docs/8.0.0/demo-moving.html
- https://adazzle.github.io/react-data-grid/docs/examples/row-reordering
- https://adazzle.github.io/react-data-grid/canary/?path=/story/demos--rows-reordering
- https://github.com/mbrn/material-table/issues/1219

**Comment by LincMitch** [comment]
How about a List as well as a Table

**Comment by Fasani** [comment]
The company I work for use Material UI, and we need this feature. 

I have two options. I guess I can have one team implement https://github.com/Autodesk/react-base-table and style it or have them complete this task for the MUI library. 

If we were to add this feature, what would the process be, is there contribution forms that need to be filled in and signed, or do we just fork and open a PR.

This page is a bit bare in terms of requirements/spec, is there more info available, or can we just implement it how we see fit for our needs.

Sorry if these topics are covered somewhere. Please point me to a link if so. Thank you.

**Comment by oliviertassinari** [comment]
@Fasani Hey, thanks for the interest in contributing features into the library. This case has come up a couple of times in the past. For Pro feature, what we generally propose is as follow:

- if the changes gets in, we provide licenses free of charges.
- in order to accept the changes, we need to sign a copyright assignment agreement.

You can sent an email at x@material-ui.com if you want to continue the conversation.

**Comment by Fasani** [comment]
Thanks for getting back to me.

I noticed this is on DataGrid and perhaps I would want it on Table. Does that make a difference?

Likely I will check out the code and start considering the change. If it makes sense I can send it back to you.

**Comment by flaviendelangle** [comment]
I doubt it will ever implemented in the Table (https://material-ui.com/components/tables/) component.
What would stop you to migrate for Table to DataTable ?

**Comment by Fasani** [comment]
I ended up going with option A, thanks for the replies.

**Comment by gyudisanyi** [comment]
Hello!
Any news on this feature would be greatly appreciated!
Thank you - also for the great job you are doing, <3 mat-ui.

**Comment by camnealie** [comment]
I note there's a feature request (mentioned in OP) for draggable rows to be implemented in table, has anyone made any progress for this in datagrid / datagrid pro? Would be nice!

**Comment by flaviendelangle** [comment]
Not yet no
We are focusing on other features that seems to be wanted by more people.
But I think the row reorder should not be a major pain to implement when we will have the time for it.

**Comment by trentschnee** [comment]
Cant wait for this feature!

**Comment by paguerrie** [comment]
This feature would convince me to upgrade to the pro plan.

**Comment by marc-polizzi** [comment]
Looking forward for this feature as well.

**Comment by david-ic3** [comment]
This one would be nice

**Comment by Snosii** [comment]
any idea, whether this feature would be available in Q1 this year?

**Comment by adasko86** [comment]
Our company is also waiting for this feature. Do you have any plans for that?

**Comment by InnocentWahome** [comment]
This would work really well for my project, I have so many rows on particular tables, re-ordering them would make analysis much easier

**Comment by oriolcx** [comment]
We have a couple of licenses and we would also be very interested in this feature.

**Comment by Mistes974** [comment]
Hi ! Why is it taking so long to merge ? We are looking forward to this feature.

**Comment by DanailH** [comment]
I apologize for the delay. I will resume the work on it today and aim to have it in this week's release.

**Comment by mohamedsendii** [comment]
yes for example to implement a hierarchy of an organization in a Datagrid tree data it will be practical to do the drag and drop of rows in order to modify the position of the nodes in this organization without going through a form, I hope that you create a new issue "Row reorder for Row Grouping and Tree Data"

**Comment by gustavozirbes** [comment]
> For now it does not If there is enough traction, we might implement it later with callbacks to update your data accordingly.
> 
> @DanailH maybe we could create a new issue "Row reorder for Row Grouping and Tree Data", put it "waiting for upvote" and link it in the doc.

Hi @flaviendelangle. Is there a clear path of what are the priorities for the data grid? I feel that things are being added without much planning in terms of closing the gaps.

For instance: the tree functionality was added but without support for lazy load, which makes it pretty hard to be used in enterprise applications. Now we have the row reorder, but it does not work with tree data.

Please don't take me wrong, I really like working with MUI components, but it is been hard to use these new features where one thing does not (fully) work with another. When I look at the roadmap, I can't see these gaps getting closed.

**Comment by flaviendelangle** [comment]
For the row reorder, I think the usage with a tree structure is far from being the majority of the use cases.
So we decided that it was better to launch the feature for all the people not using a tree structure rather than delaying.
With that being said, I agree that it is easy to forget to "close the gaps" as you say.

For the lazy loading of the tree. Is the [custom example](https://mui.com/x/react-data-grid/tree-data/#children-lazy-loading) enough for you ?
And if not, what is missing ?
We will have to implement it correctly on our side at some point. But it requires to rework our entire internal data model and we decided to go for others features before that felt easier to do relative to the community traction.
